rclpy/src/rclpy/_rclpy.c

```diff
@@ -1599,6 +1599,72 @@ rclpy_publish(PyObject * Py_UNUSED(self), PyObject * args)
   Py_RETURN_NONE;
 }
 
+/// Publish a serialized message
+/**
+ * Raises ValueError if pypublisher is not a publisher capsule
+ * Raises RuntimeError if the message cannot be published
+ *
+ * \param[in] pypublisher Capsule pointing to the publisher
+ * \param[in] pymsg serialized message to send
+ * \return NULL
+ */
+
+static PyObject *
+rclpy_publish_raw(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pypublisher;
+  PyBytesObject * pymsg;
+  char * message;
+  Py_ssize_t pymsg_size;
+
+  if (!PyArg_ParseTuple(args, "OO", &pypublisher, &pymsg)) {
+    return NULL;
+  }
+
+  int err = PyBytes_AsStringAndSize(pymsg, &message, &pymsg_size);
+
+  if (err == -1) {
+    PyErr_Format(RCLError, "Failed to fetch buffer contents: %d", err);
+    return NULL;
+  }
+
+  rcl_serialized_message_t msg = rmw_get_zero_initialized_serialized_message();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_ret_t ret = rmw_serialized_message_init(&msg, (size_t)pymsg_size, &allocator);
+
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(
+      RCLError,
+      "Failed to initialize message: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
```

Comment:
I don't think we should worry about finalizing the message if it was not successfully initialized. We can remove this line and the related if-block below.